### ISSUE TITLE
Fullscreen is not switchable.

### ISSUE THREAD

**Issue by Chaosus** [MonoGame/MonoGame#2495]
I think this issue is important and must be fixed in next release. IsFullscreen in GraphicsDeviceManager is not worked correctly, nor ToggleFullscreen. For example - developer cannot switch fullscreen through game menu now. The better way of implementing may be screen scale - "Fullscreen(windowed)" in modern games - where driver is not aquired his internal fullscreen state and alt+tab is really fast(but overall it looks like real fullscreen)

**Comment by awesomeProgrammer** [comment]
Hello
I've begun to programming with monogame for about a week ago, came to chapther 3 of my book, where I found the full-screen-bug, which have let me here.

I was about to drop monoGame and would have gone back to Löve2D, when I found this tutorial.

http://rbwhitaker.wikidot.com/changing-the-window-size

- full screen (called inside of the update function)
Graphics.PreferredBackBufferWidth = GraphicsDevice.DisplayMode.Width;
Graphics.PreferredBackBufferHeight = GraphicsDevice.DisplayMode.Height;
Graphics.IsFullScreen = true;
Graphics.ApplyChanges();

I tried it out and got a full screen, but then I couldn't go back to window-mode.

Then when I played around with the values, I discovered if you set Graphics.IsFullScreen to true in the game1 constructor, it would do nothing to the screen on startup (only on the first frame, I think), but later I could turn the screen back to window-mode again.

        public Game1()
        {
            mGraphics = new GraphicsDeviceManager(this);
            Content.RootDirectory = "Content";

            //mGraphics.PreferredBackBufferWidth = GraphicsDevice.DisplayMode.Width;
            //mGraphics.PreferredBackBufferHeight = GraphicsDevice.DisplayMode.Height;
            mGraphics.IsFullScreen = true;
            mGraphics.ApplyChanges();

        }

        protected override void Update(GameTime gameTime)
        {
            if (InputWrapper.Buttons.Back == ButtonState.Pressed)
                this.Exit();

            if (InputWrapper.Buttons.A == ButtonState.Pressed)
            {
                if (!mGraphics.IsFullScreen)
                {
                    mGraphics.PreferredBackBufferWidth = GraphicsDevice.DisplayMode.Width;
                    mGraphics.PreferredBackBufferHeight = GraphicsDevice.DisplayMode.Height;
                    mGraphics.IsFullScreen = true;
                    mGraphics.ApplyChanges();
                }
            }

            if (InputWrapper.Buttons.B == ButtonState.Pressed)
            {
                if(mGraphics.IsFullScreen)
                {
                    mGraphics.PreferredBackBufferWidth = kWindowWidth;
                    mGraphics.PreferredBackBufferHeight = kWindowHeight;
                    mGraphics.IsFullScreen = false;
                    mGraphics.ApplyChanges();
                }
            }
        }

when you toggle between the 2 modes, there are some flashes for 2 seconds, but now I got it to work.

I hope that this can help others getting past this problem.

(I have yet to do any advants testing with this system, because I don't know how XD)

**Comment by Shqrdx** [comment]
@awesomeProgrammer - try use WindowsDX instead WindowsGL platform

**Comment by Shqrdx** [comment]
btw You should not use graphics.ApplyChanges(); in game constructor.

**Comment by awesomeProgrammer** [comment]
I used the "Windows OpenGL" project option

**Comment by awesomeProgrammer** [comment]
Just tested the same code in the "Windows" project option.

it just flashes between a black screen and normal full screen, forever.

**Comment by awesomeProgrammer** [comment]
removed these 2 lines from the constructor.

Graphics.IsFullScreen = true;
Graphics.ApplyChanges();

It starts in a normal window mode.

When I use the same button functions, it works as it did before, but whenever I'm in full screen, it just flashes between a black screen and the full game screen.

so it can toggle between the 2 modes, but full-screen-mode is shit.

(that means that we all need to work in OpenGL from now on)

**Comment by Shqrdx** [comment]
OpenGL for windows is bugged.. 
try this code for Update for Windows(without GL) project

        private KeyboardState lastKS;

        /// <summary>
        /// Allows the game to run logic such as updating the world,
        /// checking for collisions, gathering input, and playing audio.
        /// </summary>
        /// <param name="gameTime">Provides a snapshot of timing values.</param>
        protected override void Update(GameTime gameTime)
        {
            KeyboardState ks = Keyboard.GetState();

            if (lastKS.IsKeyUp(Keys.Q) && ks.IsKeyDown(Keys.Q))
            {
                if (!graphics.IsFullScreen)
                {
                    graphics.PreferredBackBufferWidth = GraphicsDevice.DisplayMode.Width;
                    graphics.PreferredBackBufferHeight = GraphicsDevice.DisplayMode.Height;
                    graphics.IsFullScreen = true;
                    graphics.ApplyChanges();
                }
            }

            if (lastKS.IsKeyUp(Keys.E) && ks.IsKeyDown(Keys.E))
            {
                if (graphics.IsFullScreen)
                {
                    graphics.PreferredBackBufferWidth = GraphicsDevice.DisplayMode.Width;
                    graphics.PreferredBackBufferHeight = GraphicsDevice.DisplayMode.Height;
                    graphics.IsFullScreen = false;
                    graphics.ApplyChanges();
                }
            }

            lastKS = ks;

            base.Update(gameTime);
        }

if it continue switching - then you use wrong input class..

**Comment by awesomeProgrammer** [comment]
I fixed it now.

looked through the code with the debugger tool, there I saw that at the end of each update call, the second last line was a Graphics.ApplyChanges(), which was the problem to the flashing.

It works perfectly fine now, like there never was a problem to begin with.

now I just need someone with more MonoGame experience to test this code out, to prove that this could be a stable fix to everyone problems.
